vignettes/articles/ui-cli-conversion.Rmd

```diff
@@ -0,0 +1,442 @@
+
+```{asciicast}
+# why is this block truncated from the top in the rendered document?
+x <- glue("Your super secret password is {ui_unset()}.")
+dput(x)
+ui_info(x)
+```
+
+### cli replacements
+
+In general, we can move towards cli's native inline-markup:
+<https://cli.r-lib.org/reference/inline-markup.html>
+
+Here's the general conversion plan:
+
+-   `ui_field()` becomes `{.field blah}`. In `usethis_theme()`, I tweak the `.field` style to apply single quotes if color is not available, which is what `ui_field()` has always done.
+-   `ui_value()` becomes `{.val value}`.
+-   `ui_path()` is connected to `{.path path/to/thing}`, but, as explained above, `ui_path()` also does more. Therefore, I abstracted the "path math" into an internal helper, `ui_path_impl()`, which is aliased to `pth()` for compactness. Here's a typical conversion:
+    ```{r}
+    #| eval: false
+    # using legacy functions
+    ui_done("Setting {ui_field('LazyData')} to \\
+             {ui_value('true')} in {ui_path('DESCRIPTION')}")
+    # using new cli-based ui
+    ui_bullets(c(
+      "v" = "Setting {.field LazyData} to {.val true} in {.path {pth('DESCRIPTION')}}."
+    ))
+    ```
+    It would be nice to create a custom inline class, e.g. `{.ui_path {some_path}}`, which I have done in, e.g., googledrive.
+    But it's not easy to do this *while still inheriting cli's `file:` hyperlink behaviour*, which is very desirable.
+    So that leads to the somewhat clunky, verbose pattern above, but it gives a nice result.
+-   `ui_code()` gets replaced by various inline styles, depending on what the actual goal is, such as:
+    -   `{.code some_text}`
+    -   `{.arg some_argument}`
+    -   `{.cls some_class}`
+    -   `{.fun devtools::build_readme}`
+    -   `{.help some_function}`
+    -   `{.run usethis::usethis_function()}`
+    -   `{.topic some_topic}`
+    -   `{.var some_variable}`
+- `ui_unset()` is replaced by `ui_special()`, which you'll see more of below. Currently the intended grey color doesn't show up when I render this document using solarized-dark and so far I can't get to the bottom of that :( Why isn't it the same grey as "[Copied to clipboard]" in `ui_code_snippet()`, which does work?
+
+## Conditions
+
+I'm moving from `ui_stop()`:
+
+```{r}
+#| eval: false
+ui_stop <- function(x, .envir = parent.frame()) {
+  x <- glue_collapse(x, "\n")
+  x <- glue(x, .envir = .envir)
+
+  cnd <- structure(
+    class = c("usethis_error", "error", "condition"),
+    list(message = x)
+  )
+
+  stop(cnd)
+}
+```
+
+to `ui_abort()`:
+
+```{r}
+#| eval: false
+ui_abort <- function(message, ..., class = NULL, .envir = parent.frame()) {
+  cli::cli_div(theme = usethis_theme())
+  # bullet naming gymnastics, see below
+  cli::cli_abort(
+    message,
+    class = c(class, "usethis_error"),
+    .envir = .envir,
+    ...
+  )
+}
+```
+
+The main point of `ui_abort()` is to use to `cli_abort()` (and to continue applying the `"usethis_error"` class).
```

Comment:
I assume we mostly use that error class for tests?

Reply:
Yeah.